jina/proto/jina.proto

```diff
@@ -109,6 +101,8 @@ message Document {
 
     // the embedding array of this doc
     NdArray embedding = 19;
+
+    Relevance score = 20;
```

Comment:
should `score` and `matches`  be `either-or`? I guess a Score would be assigned to a `retrieved` document, and the `matches` would be added to the `query document`?

Reply:
setting `oneof` on `matches` and `score` will make the following pattern unimplementable.

```
doc1 --matches-- doc3 (with score ref_id=1), doc4 (with score ref_id=1), doc7 (with score ref_id=1)
                    |
                    | --- matches -- doc6 (with score ref_id=3), doc9 (with score ref_id=3), doc10 (with score ref_id=3)
```

Reply:
Is it a pattern that can happen? or it is for the sake of full flexibility?

Reply:
In theory, `doc1` should come from a `query set`, and `doc3, doc4 and doc7` from the `index set`. This bifurcation of `matches` to me would have a different `meaning`. How a `doc` from the `index set` has matches `with others`? Wouldn't that imply something like a kNN graph?

Reply:
This design:
- improves the consistency of recursive structure introduced in #652 
- allows advanced training/search paradigm that requires a proximity graph/nearest-neighbor graph (which is quite common in ML). It opens the door to graph learning.

Reply:
understood ;)

Reply:
That being said, most of our examples can be implemented with just one level of recursion.

Reply:
Yes, I see it more as a potential enabler for different features in the future.